File: trip-server/src/controllers/conversation.controller.ts
import { Request, Response } from 'express'
import * as conversationService from '../services/conversationService'
import { parseIntParam, isInvalidParamError } from '../utils/params'

export const list = async (req: Request, res: Response) => {
  if (!req.user) return res.status(401).json({ code: 401, error: '未登录' })
  const page = Number(req.query.page) || 1
  const pageSize = Number(req.query.pageSize) || 20
  try {
    const result = await conversationService.listConversations(req.user.userId, page, pageSize)
    return res.json({ code: 200, data: result })
  } catch (e) {
    return res.status(500).json({ code: 500, error: '获取对话列表失败' })
  }
}

export const detail = async (req: Request, res: Response) => {
  if (!req.user) return res.status(401).json({ code: 401, error: '未登录' })
  try {
    const id = parseIntParam(req.params.id, 'id')!
    const conv = await conversationService.getConversationDetail(id, req.user.userId)
    if (!conv) return res.status(404).json({ code: 404, error: '对话不存在' })
    return res.json({ code: 200, data: conv })
  } catch (e) {
    if (isInvalidParamError(e)) return res.status(400).json({ code: 400, error: e.message })
    return res.status(500).json({ code: 500, error: '获取对话详情失败' })
  }
}

export const remove = async (req: Request, res: Response) => {
  if (!req.user) return res.status(401).json({ code: 401, error: '未登录' })
  try {
    const id = parseIntParam(req.params.id, 'id')!
    await conversationService.deleteConversation(id, req.user.userId)
    return res.json({ code: 200, message: '删除成功' })
  } catch (e) {
    if (isInvalidParamError(e)) return res.status(400).json({ code: 400, error: e.message })
    const msg = e instanceof Error ? e.message : '删除失败'
    return res.status(400).json({ code: 400, error: msg })
  }
}

File: trip-backend-java/src/main/java/com/trip/backend/web/controller/ConversationController.java
package com.trip.backend.web.controller;

import com.trip.backend.domain.entity.Message;
import com.trip.backend.service.ConversationService;
import com.trip.backend.utils.AppException;
import com.trip.backend.web.handler.FormatResolver;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

/**
 * 会话控制器（对应 Python routers/conversation.py）
 * 4 个端点：
 * - GET /api/conversations（分页）
 * - GET /api/conversations/{id}（详情 + 消息）
 * - POST /api/conversations
 * - DELETE /api/conversations/{id}
 */
@RestController
@RequestMapping("/api/conversations")
public class ConversationController {

    private final ConversationService conversationService;

    public ConversationController(ConversationService conversationService) {
        this.conversationService = conversationService;
    }

    /**
     * GET /api/conversations
     */
    @GetMapping
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<Map<String, Object>> getConversations(
            @RequestAttribute("userId") Long userId,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int pageSize) {
        Page<com.trip.backend.domain.entity.Conversation> conversations =
            conversationService.getConversations(userId, page, pageSize);

        Map<String, Object> data = Map.of(
            "items", conversations.getContent(),
            "total", conversations.getTotalElements(),
            "page", page,
            "pageSize", pageSize
        );

        return ResponseEntity.ok(Map.of(
            "code", 200, "data", data, "message", null, "error", null
        ));
    }

    /**
     * GET /api/conversations/{id}
     */
    @GetMapping("/{id}")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<Map<String, Object>> getConversation(
            @RequestAttribute("userId") Long userId,
            @PathVariable Long id) {
        ConversationService.ConversationWithMessages cwm =
            conversationService.getConversation(userId, id);

        // 转换 Message 为 snake_case（匹配 Python 返回格式）
        List<Map<String, Object>> messages = cwm.messages().stream()
            .map(msg -> Map.of(
                "id", msg.getId(),
                "role", msg.getRole(),
                "content", msg.getContent(),
                "metadata", msg.getMetadata(),
                "created_at", msg.getCreatedAt().toString()
            ))
            .toList();

        Map<String, Object> data = Map.of(
            "id", cwm.conversation().getId(),
            "title", cwm.conversation().getTitle(),
            "messages", messages
        );

        return ResponseEntity.ok(Map.of(
            "code", 200, "data", data, "message", null, "error", null
        ));
    }

    /**
     * POST /api/conversations
     */
    @PostMapping
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<Map<String, Object>> createConversation(
            @RequestAttribute("userId") Long userId,
            @Valid @RequestBody(required = false) CreateRequest request) {
        String title = request != null ? request.title() : null;
        var conversation = conversationService.createConversation(userId, title);

        Map<String, Object> data = Map.of(
            "id", conversation.getId(),
            "title", conversation.getTitle()
        );

        return ResponseEntity.status(HttpStatus.CREATED).body(Map.of(
            "code", 201, "data", data, "message", null, "error", null
        ));
    }

    /**
     * DELETE /api/conversations/{id}
     */
    @DeleteMapping("/{id}")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<Map<String, Object>> deleteConversation(
            @RequestAttribute("userId") Long userId,
            @PathVariable Long id) {
        conversationService.deleteConversation(userId, id);
        return ResponseEntity.noContent().build();
    }

    public record CreateRequest(@NotBlank String title) {}
}
